app/src/main/java/io/appium/uiautomator2/utils/UiExpressionParser.java

```diff
@@ -0,0 +1,299 @@
+                            break;
+                    }
+                }
+            currentIndex++;
+        } while (!parenthesesStack.empty() && currentIndex < expression.length());
+
+        if (!parenthesesStack.isEmpty()) {
+            throw new UiSelectorSyntaxException("unclosed paren in expression:" + expression);
+        }
+
+        final String argument = expression.substring(startIndex + 1, currentIndex - 1);
+        if (!argument.isEmpty()) {
+            arguments.add(argument.trim());
+        }
+
+        expression = expression.substring(currentIndex);
+        return arguments;
+    }
+
+    /**
+     * consume [a-z]* then an open paren, this is our methodName
+     * consume .* and count open/close parens until the original open paren is close, this is
+     * our
+     * argument
+     */
+    protected <V> V consumeMethodCall() throws UiSelectorSyntaxException,
+            UiObjectNotFoundException {
+        final String methodName = consumeMethodName();
+        final List<String> arguments = consumeMethodParameters();
+        final Pair<Method, List<Object>> methodWithArgument = findMethod(methodName, arguments);
+        return invokeMethod(target, methodWithArgument.first, methodWithArgument.second);
+    }
+
+    protected Pair<Method, List<Object>> findMethod(String methodName, List<String> arguments)
+            throws UiSelectorSyntaxException, UiObjectNotFoundException {
+        final List<Method> candidates = new ArrayList<>();
+        for (final Method method : clazz.getDeclaredMethods()) {
+            if (method.getName().equals(methodName)) {
+                candidates.add(method);
+            }
+        }
+
+        if (candidates.isEmpty()) {
+            throw new UiSelectorSyntaxException(String.format("%s has no `%s` method",
+                    getTarget().getClass().getSimpleName(), methodName));
+        }
+
+        UiSelectorSyntaxException exThrown = null;
+        for (final Method method : candidates) {
+            try {
+                final Type[] parameterTypes = method.getGenericParameterTypes();
+                final List<Object> args = coerceArgsToTypes(parameterTypes, arguments);
+                return new Pair<>(method, args);
+            } catch (UiSelectorSyntaxException e) {
+                exThrown = e;
+            }
+        }
+
+        final String errorMsg = "`%s` doesn't have suitable method `%s` with arguments %s" +
+                (exThrown != null ? ": " + exThrown.getMessage() : "");
+
+        throw new UiSelectorSyntaxException(String.format(errorMsg, clazz.getSimpleName(),
+                methodName, arguments), exThrown);
+    }
+
+    private Pair<Constructor, List<Object>> findConstructor(List<String> arguments) throws
+            UiSelectorSyntaxException, UiObjectNotFoundException {
+        UiSelectorSyntaxException exThrown = null;
+        for (final Constructor constructor : clazz.getConstructors()) {
```

Comment:
this block of code seems to be duplicated

Reply:
There is no common superclass or interface for `Constructor` and `Method` classes in API<26.

Reply:
Would it be possible to create a custom interface with `getGenericParameterTypes` method and use forced typecast or the compiler won't like it?

Reply:
Something like this?
```java
    class A {
        public void foo() {
        }
    }
    interface B {
        void foo();
    }
    B b = (B) new A();
```
No, you can't do that tricks in java world.